packages/sst/support/custom-resources/s3-uploader.ts

```diff
@@ -0,0 +1,189 @@
+            destinationBucketName,
+            fileOptions,
+            replaceValues,
+          } satisfies ProcessorEvent),
+        })
+      );
+    })
+  );
+}
+
+async function purgeOldFiles(props: Props) {
+  if (!props.filenames) return;
+
+  // Get all uploaded files
+  const s3Object = await s3.send(
+    new GetObjectCommand({
+      Bucket: props.filenames.bucketName,
+      Key: props.filenames.objectKey,
+    })
+  );
+  const content = await s3Object.Body?.transformToString();
+  if (!content) throw new Error("File content is empty");
+  const uploadedFiles = content.split("\n");
+
+  // Get all files in destination bucket
+  const files = await s3.send(
+    new ListObjectsV2Command({
+      Bucket: props.destinationBucketName,
+    })
+  );
+  const bucketFiles = files.Contents?.map((file) => file.Key);
+  if (!bucketFiles) throw new Error("No files found in destination bucket");
+
+  // Remove files that are not in the uploaded files
+  for (const bucketFile of bucketFiles) {
+    if (!bucketFile) continue;
+    if (uploadedFiles.includes(bucketFile)) continue;
+
+    await s3.send(
+      new DeleteObjectCommand({
+        Bucket: props.destinationBucketName,
+        Key: bucketFile,
+      })
+    );
+  }
+}
+
+export async function batchProcessor(props: ProcessorEvent) {
+  const { source, destinationBucketName, fileOptions, replaceValues } = props;
+
+  // Create a temporary working directory
+  const contentsDir = path.join("/", "tmp", "contents");
+  log({ contentsDir });
+  fs.rmSync(contentsDir, { recursive: true, force: true });
+  fs.mkdirSync(contentsDir, { recursive: true });
+
+  // Download file and extract to "contents"
+  const response = await s3.send(
+    new GetObjectCommand({
+      Bucket: source.bucketName,
+      Key: source.objectKey,
+    })
+  );
+  const stream = response.Body as Readable;
+  const zip = new AdmZip(
+    await new Promise((resolve, reject) => {
+      const chunks: Buffer[] = [];
+      stream.on("data", (chunk) => chunks.push(chunk));
+      stream.on("error", reject);
+      stream.on("end", () => resolve(Buffer.concat(chunks)));
+    })
+  );
+  zip.extractAllTo(contentsDir, true);
+
+  // Replace values in files
+  for (const replaceValue of replaceValues) {
+    log("> pattern", replaceValue.files);
+    const files = globSync(replaceValue.files, {
+      cwd: contentsDir,
+      nodir: true,
+    });
+    for (const file of files) {
+      log("> > file", file);
+      const content = fs
+        .readFileSync(file, "utf8")
+        .replaceAll(replaceValue.search, replaceValue.replace);
+      fs.writeFileSync(file, content);
+    }
+  }
+
+  // Upload to S3
+  const filesUploaded: string[] = [];
+  for (const fileOption of fileOptions) {
+    const files = globSync(fileOption.files, {
+      cwd: contentsDir,
+      nodir: true,
+      ignore: fileOption.ignore,
+    }).filter((file) => !filesUploaded.includes(file));
+    for (const file of files) {
+      await s3.send(
```

Comment:
I suspect between the file I/O and whatever networking is involved in the `s3.send`, this operation would probably greatly benefit from utilizing a bulkhead with maybe 3-5 concurrent requests and changing to an async `readFile`.